src/lib/provable/gadgets/blake2b.ts

```diff
@@ -0,0 +1,189 @@
+// https://www.blake2.net/blake2.pdf
+import { UInt64, UInt8 } from '../int.js';
+import { FlexibleBytes } from '../bytes.js';
+import { Bytes } from '../wrapped-classes.js';
+
+export { BLAKE2B };
+
+const BLAKE2BConstants = {
+  IV: [
+    0x6a09e667f3bcc908n,
+    0xbb67ae8584caa73bn,
+    0x3c6ef372fe94f82bn,
+    0xa54ff53a5f1d36f1n,
+    0x510e527fade682d1n,
+    0x9b05688c2b3e6c1fn,
+    0x1f83d9abfb41bd6bn,
+    0x5be0cd19137e2179n,
+  ],
+
+  SIGMA: [
+    [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
+    [14, 10, 4, 8, 9, 15, 13, 6, 1, 12, 0, 2, 11, 7, 5, 3],
+    [11, 8, 12, 0, 5, 2, 15, 13, 10, 14, 3, 6, 7, 1, 9, 4],
+    [7, 9, 3, 1, 13, 12, 11, 14, 2, 6, 5, 10, 4, 0, 15, 8],
+    [9, 0, 5, 7, 2, 4, 10, 15, 14, 1, 11, 12, 6, 8, 3, 13],
+    [2, 12, 6, 10, 0, 11, 8, 3, 4, 13, 7, 5, 15, 14, 1, 9],
+    [12, 5, 1, 15, 14, 13, 4, 10, 0, 7, 6, 3, 9, 2, 8, 11],
+    [13, 11, 7, 14, 12, 1, 3, 9, 5, 0, 15, 4, 8, 6, 2, 10],
+    [6, 15, 14, 9, 11, 3, 0, 8, 12, 2, 13, 7, 1, 4, 10, 5],
+    [10, 2, 8, 4, 7, 6, 1, 5, 15, 11, 9, 14, 3, 12, 13, 0],
+  ],
+};
+
+const BLAKE2B = {
+  hash(data: FlexibleBytes, digest_length = 64) {
+    const state = initialize(digest_length);
+    update(state, Bytes.from(data).bytes);
+    const out = final(state);
+    return Bytes.from(out);
+  },
+  get IV() {
+    return BLAKE2BConstants.IV.map((x) => UInt64.from(x));
+  },
+};
+
+function initialize(outlen: number): {
+  h: UInt64[];
+  t: UInt64[];
+  f: UInt64[];
+  buf: UInt8[];
+  buflen: number;
+  outlen: number;
+} {
+  const h = BLAKE2B.IV.slice(); // shallow copy IV to h
+  h[0] = UInt64.from(0x01010000).xor(UInt64.from(outlen)).xor(h[0]);
+  return {
+    h,
+    t: [UInt64.zero, UInt64.zero],
+    f: [UInt64.zero, UInt64.zero],
+    buf: [],
+    buflen: 0,
+    outlen,
+  };
+}
+
+function G(
+  v: UInt64[],
+  a: number,
+  b: number,
+  c: number,
+  d: number,
+  x: UInt64,
+  y: UInt64
+) {
+  v[a] = v[a].addMod64(v[b].addMod64(x));
```

Comment:
I think we could save a good chunk of constraints on some of these by using addition on field elements, ignoring the overflow and only at the very end reducing them mod64, eg

```diff
- v[a] = v[a].addMod64(v[b].addMod64(x));
 
+ const unreducedVA = v[a].value.add(v[b].value.add(x.value));
+ v[a] = UInt64.Unsafe.fromField(Gadgets.divMod64(unreducedVA, 1).remainder);
```

We can do this because `divMod64` range checks the `remainder` to 64 bit `rangeCheck64(remainder);` already, so we don't need to do this at every addition step in the chain -- this way we save at least one call to `addMod64`

Reply:
Yeah, we can do that. Thanks for pointing that out. Second argument of `divMod64` should be 64 for the new implementation of it.